platform/admin/admin-configuration-configupdater/src/main/java/org/codice/ddf/admin/configuration/ConfigurationUpdater.java

```diff
@@ -0,0 +1,322 @@
+    if (fileFromCache != null && fileFromCache.exists()) {
+      LOGGER.debug("Deleting file because config was deleted for pid [{}]", pid);
+      try {
+        Files.delete(fileFromCache.toPath());
+        SecurityLogger.audit("Removing a deleted config [{}]", fileFromCache.toPath());
+      } catch (IOException e) {
+        LOGGER.debug(
+            format("Problem deleting config file [%s]: ", fileFromCache.getAbsolutePath()), e);
+        // Synchronous with config admin, so we can report the failure to the UI this way
+        throw e;
+      }
+    }
+  }
+
+  private void writeIfNecessary(
+      String appropriatePid,
+      File dest,
+      Dictionary<String, Object> configState,
+      CachedConfigData cachedData)
+      throws IOException {
+    if (dest != null && dest.exists()) {
+      processPasswords(appropriatePid, configState, this::encryptValue);
+      if (!cachedData.equalProps(configState)) {
+        try (OutputStream outputStream = new BufferedOutputStream(new FileOutputStream(dest))) {
+          getAppropriateStrategy(dest.toPath()).write(outputStream, configState);
+        } catch (IOException e) {
+          LOGGER.debug("Writing to config file failed: ", e);
+          SecurityLogger.audit("Failure to update config file [{}]", dest.getAbsolutePath());
+          throw e;
+        }
+        SecurityLogger.audit("Updating config file [{}]", dest.toPath());
+        cachedData.setProps(configState);
+      }
+    }
+  }
+
+  /**
+   * {@link EncryptionService} is whitelisted, so we cannot add this method onto the interface and
+   * impl until a major release.
+   */
+  private String encryptValue(String plaintextValue) {
+    Matcher m = ENC_PATTERN.matcher(plaintextValue);
+    if (m.find()) {
+      LOGGER.debug("Password already encrypted");
+      return plaintextValue;
+    }
+    return format("ENC(%s)", encryptionService.encrypt(plaintextValue));
+  }
+
+  private void processPasswords(
+      String appropriatePid,
+      Dictionary<String, Object> dictionary,
+      Function<String, String> passwordTransform) {
+    ObjectClassDefinition objectClassDefinition =
+        ddfConfigAdmin.getObjectClassDefinition(appropriatePid);
+    if (objectClassDefinition == null) {
+      return;
+    }
+    Stream.of(objectClassDefinition)
+        .map(ocd -> ocd.getAttributeDefinitions(ObjectClassDefinition.ALL))
+        .flatMap(Arrays::stream)
+        .filter(ad -> ad.getType() == AttributeDefinition.PASSWORD)
+        .map(AttributeDefinition::getID)
+        .filter(id -> dictionary.get(id) instanceof String)
+        .forEach(id -> dictionary.put(id, passwordTransform.apply((String) dictionary.get(id))));
+  }
+
+  private PersistenceStrategy getAppropriateStrategy(Path path) {
+    String ext = getFileExtension(path.toString());
+    if (ext.isEmpty()) {
+      throw new IllegalArgumentException("Path has no file extension");
```

Comment:
❓ Does that mean that this would now bubble out all the way to the code that initiated an update with a felix filename that has an extension we do not support?
❗️ I am concerned about that.
✏️  We definitely a log for this ... warning?

Reply:
Log warning added. 

The directory watcher should only be looking for config file changes on `.cfg` and `.config` since we already have `.xml` in etc anyway and they get ignored. Unless that gets less restrictive this shouldn't happen. If someone changes the felix filename to anything other than null then that's an illegal state exception which is viable on this API.